### PR TITLE
Added suppport for Kafka partitioner and explicit setting of partition on record

### PR DESCRIPTION
**Pull request by merlimat**
### Motivation

 * Support setting partition id on records when publishing through Pulsar-Kafka adapter library
 * Support producer config `partitioner.class` with a custom partitioner

### CODE REVIEW

**Comment by jiazhai** [comment]
run integration tests

**Comment by merlimat** [comment]
retest this please

File: site2/docs/adaptors-kafka.md [hunk 1]
```diff
@@ -129,7 +129,7 @@ Properties:
 | Config property                         | Supported | Notes                                                                         |
 |:----------------------------------------|:----------|:------------------------------------------------------------------------------|
 | `acks`                                  | Ignored   | Durability and quorum writes are configured at the namespace level            |
-| `auto.offset.reset`			  | Yes       | Will have a default value of 'latest' if user does not give specific setting. |
+| `auto.offset.reset`		 	          | Yes       | Will have a default value of `latest` if user does not give specific setting. |
```
Comment (ivankelly) [inline site2/docs/adaptors-kafka.md]: Added superfluous whitespace.

File: pulsar-client-kafka-compat/pulsar-client-kafka/src/main/java/org/apache/kafka/clients/producer/PulsarKafkaProducer.java [hunk 2]
```diff
@@ -232,12 +243,11 @@ public void close(long timeout, TimeUnit unit) {
...
     private int buildMessage(TypedMessageBuilder<byte[]> builder, ProducerRecord<K, V> record) {
-        if (record.partition() != null) {
-            throw new UnsupportedOperationException("");
-        }
-
+        byte[] keyBytes = null;
         if (record.key() != null) {
-            builder.key(getKey(record.topic(), record.key()));
+            String key = getKey(record.topic(), record.key());
+            keyBytes = key.getBytes();
```
Comment (ivankelly) [inline pulsar-client-kafka-compat/pulsar-client-kafka/src/main/java/org/apache/kafka/clients/producer/PulsarKafkaProducer.java]: Add UTF_8

File: tests/pulsar-kafka-compat-client-test/src/test/java/org/apache/pulsar/tests/integration/compat/kafka/KafkaApiTest.java [hunk 3]
```diff
@@ -300,6 +302,133 @@ public void testPartitions() throws Exception {
...
+        producerProperties.put("partitioner.class", MyCustomPartitioner.class.getName());
+
+        @Cleanup
+        Producer<Integer, String> producer = new KafkaProducer<>(producerProperties);
+
+        Properties props = new Properties();
+        props.put("bootstrap.servers", getPlainTextServiceUrl());
+        props.put("group.id", "my-subscription-name");
+        props.put("enable.auto.commit", "true");
+        props.put("key.deserializer", StringDeserializer.class.getName());
```
Comment (ivankelly) [inline tests/pulsar-kafka-compat-client-test/src/test/java/org/apache/pulsar/tests/integration/compat/kafka/KafkaApiTest.java]: Shouldn't it be IntegerDeserializer?

Comment (merlimat) [reply tests/pulsar-kafka-compat-client-test/src/test/java/org/apache/pulsar/tests/integration/compat/kafka/KafkaApiTest.java]: Yes, I have mixed the consumer copied from a different test and it was assuming string key. Fixed.

File: site2/docs/adaptors-kafka.md (hunk 1 above)
Comment (merlimat) [reply site2/docs/adaptors-kafka.md]: I just fixed the `..` and aligned the columns in the source md text structure.

**Comment by merlimat** [comment]
@ivankelly PTAL again

File: pulsar-client-kafka-compat/pulsar-client-kafka/src/main/java/org/apache/kafka/clients/producer/PulsarKafkaProducer.java [hunk 4]
```diff
@@ -18,6 +18,7 @@
  */
 package org.apache.kafka.clients.producer;
 
+import java.nio.charset.Charset;
```
Comment (ivankelly) [inline pulsar-client-kafka-compat/pulsar-client-kafka/src/main/java/org/apache/kafka/clients/producer/PulsarKafkaProducer.java]: import static java.nio.charset.StandardCharsets.UTF_8;

File: site2/docs/adaptors-kafka.md (hunk 1 above)
Comment (ivankelly) [reply site2/docs/adaptors-kafka.md]: they're not aligned though: https://raw.githubusercontent.com/apache/pulsar/09da58bcda81cba3d6b29331e7ce6884bd81f150/site2/docs/adaptors-kafka.md

**Comment by merlimat** [comment]
retest this please

**Comment by merlimat** [comment]
run cpp tests
run java8 tests

**Comment by merlimat** [comment]
run integration tests

**Pull request merged**